Params is an object

nisevi opened the pull request:
fixes #2188

nisevi commented on app/presenters/event_instance_presenter.rb:
```diff
@@ -55,18 +55,27 @@ def video_embed_link
+  def participants_as_array(participants)
+    participants.map{ |participant| process_users(participant) }.compact
+  end
+
```
@tansaku I would like to unify those two methods above, I tried a couple of things but none of them convinced me. Do you think this way is ok, or should I change it?

tansaku replied:
I would probably hold off refactoring here without better understanding the context of these methods

nisevi replied:
This was not done in order to refactor, you will understand better if you pull the changes and run it locally. You will find that this method is receiving `participants` sometimes as an `Array` and sometimes as `ActionController::Parameters` object, so in order to handle both I had to split the code.

tansaku replied:
aha, then maybe:

```ruby 
  def map_to_users(participants)
    participants ||= ActionController::Parameters.new({})
    participants = participants.to_unsafe_h unless participants.is_a?(Array)
    participants.map{ |participant| process_users(participant) }.compact
  end
```

nisevi commented:
@tansaku this is not a refactoring ticket, and the changes I have done are not because I wanted to refactor the code, I know it can be confusing since I haven't write a proper description.

I will proceed to explain you why the changes I have done.

We used to have this code in here `app/presenters/event_instance_presenter.rb`:

```ruby
  def map_to_users(participants)
    participants ||= []

    participants.map do |participant|
      begin
        person = participant.last['person']
        user = Authentication.find_by(provider: 'gplus', uid: person['id']).try!(:user)
        next if user == host
        user || NullUser.new(person[:displayName])
      rescue NoMethodError
        Rails.logger.error "Exception at event_instance_presenter#map_to_users"
        nil
      end
    end.compact
  end
```

As you can see the method is receiving `participants`. And the first problem I had, was when I modified the factory `factories/event_instances.rb`, where we used to have:

`participants { (generate :broadcaster).merge((generate :participant)) }`

I replaced that with:

`participants { ActionController::Parameters.new((generate :broadcaster).merge(generate :participant)) }`

Since our parameters now are objects from this class `ActionController::Parameters`.

After I did this modification, these tests were failing:

```ruby
rspec ./spec/services/karma_calculator_spec.rb:40 # KarmaCalculator for existing members for members attending hangouts gives points for hangout participation
rspec ./spec/models/user_spec.rb:411 # User karma #number_hangouts_started_with_more_than_one_participant two event instances returns total number of hangouts started with more than one participant
rspec ./spec/models/user_spec.rb:404 # User karma #number_hangouts_started_with_more_than_one_participant single event instance returns total number of hangouts started with more than one participant
rspec ./spec/presenters/event_instance_presenter_spec.rb:35 # EventInstancePresenter all fields are present returns an array of participants
rspec ./spec/presenters/event_instance_presenter_spec.rb:42 # EventInstancePresenter all fields are present do not show the host in the list of participants
```

So I realized that these two tests:

```ruby
rspec ./spec/presenters/event_instance_presenter_spec.rb:35 # EventInstancePresenter all fields are present returns an array of participants
rspec ./spec/presenters/event_instance_presenter_spec.rb:42 # EventInstancePresenter all fields are present do not show the host in the list of participants

```

Were failing because they needed the `to_unsafe_h` method, since now `participants` was an instance of `ActionController::Parameters`:

```ruby
    it 'returns an array of participants' do
      participant = FactoryBot.create(:user, gplus: hangout.participants.to_unsafe_h.first.last['person']['id'])

      expect(presenter.participants.count).to eq(2)
      expect(presenter.participants.first).to eq(participant)
    end

    it 'do not show the host in the list of participants' do
      FactoryBot.create(:user, gplus: hangout.participants.to_unsafe_h.first.last['person']['id'])
      expect(presenter.participants).not_to include(hangout.user)
    end
```

And the method `map_users` here `app/presenters/event_instance_presenter.rb`, needed to change its default value, which it was an `Array`. So I changed it to an instance of the [class](http://api.rubyonrails.org/classes/ActionController/Parameters.html) `ActionController::Parameters`:

```ruby
  def map_to_users(participants)
    participants ||= ActionController::Parameters::new({})

    participants.map do |participant|
      begin
        person = participant.last['person']
        user = Authentication.find_by(provider: 'gplus', uid: person['id']).try!(:user)
        next if user == host
        user || NullUser.new(person[:displayName])
      rescue NoMethodError
        Rails.logger.error "Exception at event_instance_presenter#map_to_users"
        nil
      end
    end.compact
  end
```
I did this based on the assumption that `participants` on the factory that is being created is now an instance of `ActionController::Parameters`.

After I did those changes I kept having the errors on the tests, more specifically:

```ruby
bers ./spec/presenters/event_instance_presenter_spec.rb:42
Run options: include {:locations=>{"./spec/presenters/event_instance_presenter_spec.rb"=>[42]}}

Randomized with seed 54978
F

Failures:

  1) EventInstancePresenter all fields are present do not show the host in the list of participants
     Failure/Error:
       participants.map do |participant|
         begin
           person = participant.last['person']
           user = Authentication.find_by(provider: 'gplus', uid: person['id']).try!(:user)
           next if user == host
           user || NullUser.new(person[:displayName])
         rescue NoMethodError
           Rails.logger.error "Exception at event_instance_presenter#map_to_users"
           nil
         end
     
     NoMethodError:
       undefined method `map' for #<ActionController::Parameters:0x00559f7ed0bcb8>
       Did you mean?  tap
     # ./app/presenters/event_instance_presenter.rb:60:in `map_to_users'
     # ./app/presenters/event_instance_presenter.rb:29:in `participants'
     # ./spec/presenters/event_instance_presenter_spec.rb:44:in `block (3 levels) in <top (required)>'

Finished in 0.95059 seconds (files took 9.72 seconds to load)
1 example, 1 failure

Failed examples:

rspec ./spec/presenters/event_instance_presenter_spec.rb:42 # EventInstancePresenter all fields are present do not show the host in the list of participants

Randomized with seed 54978

Coverage report generated for RSpec to /home/nisevi/documents/agileventures/WebsiteOne/coverage. 368 / 2676 LOC (13.75%) covered.
```
So I thought, right!!! the instance of `ActionController::Parameters` doesn't have a `map` method!! silly me :P . So I changed the code on `app/presenters/event_instance_presenter.rb`:

`participants.map do |participant|`

by adding the `to_unsafe_h`:

`participants.to_unsafe_h.map do |participant|`

And that fixed the error for these tests:

![image](https://user-images.githubusercontent.com/3024789/37241837-22979aec-2460-11e8-9420-8ef51866ba36.png)

But!! There were still failing tests:

```ruby
rspec ./spec/models/user_spec.rb:404 # User karma #number_hangouts_started_with_more_than_one_participant single event instance returns total number of hangouts started with more than one participant
rspec ./spec/models/user_spec.rb:411 # User karma #number_hangouts_started_with_more_than_one_participant two event instances returns total number of hangouts started with more than one participant
rspec ./spec/presenters/event_instance_presenter_spec.rb:98 # EventInstancePresenter some fields are missing returns an array with nullUser if participant gplus_id is not found
rspec ./spec/presenters/event_instance_presenter_spec.rb:103 # EventInstancePresenter some fields are missing don't throw an exception when have nil person at participants
rspec ./spec/services/karma_calculator_spec.rb:40 # KarmaCalculator for existing members for members attending hangouts gives points for hangout participation
```

And running one of them:

```ruby
bers ./spec/presenters/event_instance_presenter_spec.rb:98
Run options: include {:locations=>{"./spec/presenters/event_instance_presenter_spec.rb"=>[98]}}

Randomized with seed 28476
F

Failures:

  1) EventInstancePresenter some fields are missing returns an array with nullUser if participant gplus_id is not found
     Failure/Error:
       participants.to_unsafe_h.map do |participant|
         begin
           person = participant.last['person']
           user = Authentication.find_by(provider: 'gplus', uid: person['id']).try!(:user)
           next if user == host
           user || NullUser.new(person[:displayName])
         rescue NoMethodError
           Rails.logger.error "Exception at event_instance_presenter#map_to_users"
           nil
         end
     
     NoMethodError:
       undefined method `to_unsafe_h' for #<Array:0x0055879406c978>
     # ./app/presenters/event_instance_presenter.rb:60:in `map_to_users'
     # ./app/presenters/event_instance_presenter.rb:29:in `participants'
     # ./spec/presenters/event_instance_presenter_spec.rb:100:in `block (3 levels) in <top (required)>'

Finished in 0.45652 seconds (files took 10.5 seconds to load)
1 example, 1 failure

Failed examples:

rspec ./spec/presenters/event_instance_presenter_spec.rb:98 # EventInstancePresenter some fields are missing returns an array with nullUser if participant gplus_id is not found

Randomized with seed 28476

Coverage report generated for RSpec to /home/nisevi/documents/agileventures/WebsiteOne/coverage. 360 / 2677 LOC (13.45%) covered.
```

I realize that now I was trying to apply `to_unsafe_h` to an `Array` and an object from the class `Array` doesn't have the method `to_unsafe_h`. So I figured out that I needed to split our logic, since before my changes this method was receiving `Hash` or an `Array` object. And now the method is receiving an `Array` or an `ActionController::Parameters` object and since `Array` doesn't have the method `to_unsafe_h`. I did this change:

```ruby
  def map_to_users(participants)
    participants ||= ActionController::Parameters.new({})
    return participants_as_array(participants) if participants.is_a?(Array)
    participants_as_parameters(participants)
  end

  def participants_as_parameters(participants)
    participants.to_unsafe_h.map{ |participant| process_users(participant) }.compact
  end

  def participants_as_array(participants)
    participants.map{ |participant| process_users(participant) }.compact
  end

  def process_users(participant)
    begin
      person = participant.last['person']
      user = Authentication.find_by(provider: 'gplus', uid: person['id']).try!(:user)
      user || NullUser.new(person[:displayName]) if user != host
    rescue NoMethodError
      Rails.logger.error "Exception at event_instance_presenter#map_to_users"
      nil
    end
  end
```

Which it makes our tests pass. As you see these changes are not coming from a desire for refactoring the code, but from a more complicated logic. Of course I might be wrong in the approach I have chosen, but this is not a refactor.

After those changes, I also realize that these two methods:

```
  def participants_as_parameters(participants)
    participants.to_unsafe_h.map{ |participant| process_users(participant) }.compact
  end

  def participants_as_array(participants)
    participants.map{ |participant| process_users(participant) }.compact
  end
```

Were pretty much the same, and I asked you to take a look for unifying them, because if I'm able to unify them I won't need to create the the three extra methods that I have created :+1: .

I hope this helps you understand the process I've been through when doing these changes :+1:

nisevi commented:
@tansaku thank you for adding the `fixes #2188` to the description of the issue :+1:

tansaku commented:
@nisevi gosh - thanks for the detailed description - I hadn't understood what was happening there ...

nisevi commented:
@tansaku right!! that maybe is what is happening, we are still working with data on our tests that matches rails 4 expectations and not rails 5. I will take a look to all tests that are involved in this :+1: .

nisevi commented:
@tansaku I have opened an issue on code-climate since the test coverage is dropping badly and I don't know why :+1:

nisevi commented:
@tansaku 

![image](https://user-images.githubusercontent.com/3024789/37302891-191594fa-262d-11e8-8aff-7bc00eaaeb1d.png)
